Foundation/NSNotification.swift

```diff
@@ -57,33 +58,43 @@ private let _defaultCenter: NSNotificationCenter = NSNotificationCenter()
 
 public class NSNotificationCenter : NSObject {
     
-    private var observers: [NSNotificationReceiver]
-    
-    public required override init() {
-        observers = [NSNotificationReceiver]()
-    }
+    private var _observers = [NSNotificationReceiver]()
+    private let _observersLock = NSLock()
     
     public class func defaultCenter() -> NSNotificationCenter {
         return _defaultCenter
     }
     
     public func postNotification(notification: NSNotification) {
+
         let name = notification.name
         let sender = notification.object
-        
-        let sendTo = observers.filter { observer in
+
+        _observersLock.lock()
+        let sendTo = _observers.filter { observer in
             let sameName = (observer.name == nil || observer.name == name)
             let sameSender = (observer.sender == nil || observer.sender === sender)
             
             return sameSender && sameName
         }
+        _observersLock.unlock()
         
         for observer in sendTo {
             guard let block = observer.block else {
                 continue
             }
             
-            block(notification)
+            if let queue = observer.queue {
```

Comment:
We should also check if we're already running in this queue, using `NSOperationQueue.currentQueue()`, and in that case we need to simply invoke the block otherwise we can enter in a deadlock if the queue is configured to be serial. This scenario may be unlikely but is in fact possible so I think we should have this extra validation.